arangod/Aql/IResearchViewExecutor.h

```diff
@@ -333,15 +341,169 @@ class IResearchViewExecutor {
+
+template<bool ordered>
+struct IResearchViewExecutorTraits<IResearchViewExecutor<ordered>> {
+  using IndexBufferValueType = LocalDocumentId;
+  static constexpr const bool Ordered = ordered;
+};
+
+template <bool ordered>
+class IResearchViewMergeExecutor : public IResearchViewExecutorBase<IResearchViewMergeExecutor<ordered>> {
+ public:
+  using Base = IResearchViewExecutorBase<IResearchViewMergeExecutor<ordered>>;
+  using Fetcher = typename Base::Fetcher;
+  using Infos = typename Base::Infos;
+
+  static constexpr const bool Ordered = ordered;
+
+  IResearchViewMergeExecutor(IResearchViewMergeExecutor&&) = default;
+  IResearchViewMergeExecutor(Fetcher& fetcher, Infos&);
+
+ private:
+  friend Base;
+
+  using ReadContext = typename Base::ReadContext;
+  using IndexReadBufferEntry = typename Base::IndexReadBufferEntry;
+
+  struct Segment {
+    Segment(irs::doc_iterator::ptr&& docs,
+            irs::score const& score,
+            LogicalCollection const& collection,
+            irs::columnstore_reader::values_reader_f&& sortReader,
+            irs::columnstore_reader::values_reader_f&& pkReader) noexcept
+      : docs(std::move(docs)),
+        score(&score),
+        collection(&collection),
+        sortReader(std::move(sortReader)),
+        pkReader(std::move(pkReader)) {
+      TRI_ASSERT(this->docs);
+      TRI_ASSERT(this->score);
+      TRI_ASSERT(this->collection);
+      TRI_ASSERT(this->pkReader);
+    }
+    Segment(Segment const&) = delete;
+    Segment(Segment&&) = default;
+    Segment& operator=(Segment const&) = delete;
+    Segment& operator=(Segment&&) = default;
+
+    irs::doc_iterator::ptr docs;
+    irs::score const* score{};
+    arangodb::LogicalCollection const* collection{}; // collecton associated with a segment
+    irs::bytes_ref sortValue{ irs::bytes_ref::NIL }; // sort column value
+    irs::columnstore_reader::values_reader_f sortReader; // sort column reader
+    irs::columnstore_reader::values_reader_f pkReader;  // primary key reader
+  };
+
+  class MinHeapContext {
+   public:
+    MinHeapContext(
+        iresearch::IResearchViewSort const& sort,
+        std::vector<Segment>& segments) noexcept
+      : _less(sort),
+        _segments(&segments) {
+    }
+
+    // advance
+    bool operator()(const size_t i) const {
+      assert(i < (*_segments).size());
+      auto& segment = (*_segments)[i];
+      while (segment.docs->next()) {
+        auto const doc = segment.docs->value();
+
+        if (segment.sortReader(doc, segment.sortValue)) {
+          return true;
+        }
+
+        // FIXME read pk as well
+      }
+      return false;
+    }
+
+    // compare
+    bool operator()(const size_t lhs, const size_t rhs) const {
+      assert(lhs < (*_segments).size());
+      assert(rhs < (*_segments).size());
+      return _less((*_segments)[rhs].sortValue, (*_segments)[lhs].sortValue);
+    }
+
+    iresearch::VPackComparer _less;
+    std::vector<Segment>* _segments;
```

Comment:
Wouldn't a `std::vector<Segment>&` be more convenient? The constructor gets a ref anyway...

Reply:
I would avoid having non-const reference members for object that are required to be copyable like comparators in STL, that's usually error-prone.